go/worker/sentry/grpc/worker.go

```diff
@@ -0,0 +1,234 @@
+	return func(ctx context.Context,
+		fullMethodName string,
+		req interface{}) error {
+
+		serviceName := cmnGrpc.ServiceNameFromMethod(fullMethodName)
+		if serviceName == "" {
+			g.logger.Error("error getting service name from method",
+				"method_name", fullMethodName,
+			)
+			return status.Errorf(codes.PermissionDenied, "invalid service")
+		}
+		g.RLock()
+		defer g.RUnlock()
+
+		// Get method request type.
+		methodDesc, err := cmnGrpc.GetRegisteredMethod(fullMethodName)
+		if err != nil {
+			g.logger.Error("error getting registered gRPC method",
+				"method_name", fullMethodName,
+				"err", err,
+			)
+			return status.Errorf(codes.PermissionDenied, "invalid method")
+		}
+
+		if !methodDesc.IsAccessControlled(req) {
+			// No access controll, allow.
+			return nil
+		}
+
+		p, ok := g.grpcPolicyCheckers[serviceName]
+		if !ok {
+			g.logger.Error("no policy checker defined for service",
+				"service_name", serviceName,
+			)
+			return status.Errorf(codes.PermissionDenied, "invalid service")
+		}
+
+		// Proxy unmarshals into RawMessage.
+		rawCBOR, ok := req.(*cbor.RawMessage)
+		if !ok {
+			g.logger.Error("invalid proxy request type, expected *cbor.RawMessage",
+				"request", req,
+				"request_type", fmt.Sprintf("%T", req),
+			)
+			return status.Errorf(codes.PermissionDenied, "invalid request")
+		}
+
+		// Deserialize into correct type.
+		request, err := methodDesc.UnmarshalRawMessage(rawCBOR)
+		if err != nil {
+			g.logger.Error("error unamrshaling raw request",
+				"err", err,
+				"raw", rawCBOR,
+			)
+			return status.Errorf(codes.PermissionDenied, "invalid request")
+		}
+
+		// Extract namespace.
+		namespace, err := methodDesc.ExtractNamespace(request)
+		if err != nil {
+			g.logger.Error("error extracting namespace from request",
+				"err", err,
+				"request", request,
+			)
+			return status.Errorf(codes.PermissionDenied, "invalid request")
+		}
+
+		return p.CheckAccessAllowed(ctx, accessctl.Action(fullMethodName), namespace)
+	}
+}
+
+func (g *Worker) updatePolicies(p policyAPI.ServicePolicies) {
+	g.logger.Debug("updating policies",
+		"policy", p,
+	)
+
+	g.Lock()
+	defer g.Unlock()
+	g.grpcPolicyCheckers[p.Service] = policy.NewDynamicRuntimePolicyChecker(p.Service, nil)
+	for namespace, policy := range p.AccessPolicies {
+		g.grpcPolicyCheckers[p.Service].SetAccessPolicy(policy, namespace)
+	}
+
+}
+
+func (g *Worker) worker() {
+	go func() {
+		defer close(g.quitCh)
+		defer (g.cancelCtx)()
+		<-g.grpc.Quit()
+	}()
+
+	g.policyWatcher = policyAPI.NewPolicyWatcherClient(g.conn)
+
+	// XXX: this could also be closed after getting the initial policy.
+	close(g.initCh)
+
+	pCh, pSub, err := g.policyWatcher.WatchPolicies(g.ctx)
+	if err != nil {
+		g.logger.Error("failed to watch policies",
+			"err", err,
+		)
+		panic("failed to watch policies")
+	}
+	defer pSub.Close()
+
+	// Watch Policies.
+	for {
+		select {
+		case p, ok := <-pCh:
+			if !ok {
+				g.logger.Error("WatchPolicies stream failure")
```

Comment:
probably not here, but there should be some other log message indicating the reason (i assume the error returned by the `grpc/policy/api/handlerWatchPolicies` is logged somewhere by grpc)

Reply:
ok i tested it out, so this indeed happens when the connection to the upstream node fails. This can be seen in other logs, so probably no additional logging necessary here.

Reply:
Ok then maybe just a comment saying that this happens when the connection to the upstream node fails. Also, should it automatically retry instead of just failing?

Reply:
Yeah i was thinking about this. Since in probably all cases the upstream node will be on a different host, it's not unreasonable to expect this failing from time to time. Currently what would happen is the sentry would shutdown (and probably in any reasonable deployment get automatically restarted), but we could also handle retrying here i guess.